### ISSUE TITLE
Need defensive membrane at boundary between ZCF and contract

### ISSUE THREAD

**Issue by erights** [Agoric/agoric-sdk#2497]
With one exception, all of our severe security boundaries also coincide with vat boundaries. This minimizes the semantics across those boundaries to what marshal supports, which is much narrowing than arbitrary object-to-object interaction.

The one exception is the boundary between ZCF and each contract. ZCF must be fully defensive against all possible misbehavior by the contract. We have been writing code approximately in an ocap defensive style of Jessie programming. However, we don't have the static checks to ensure that our code
   * does not leak unhardened objects
   * does not leak `this` values in callbacks
   * does full input validation of arguments

I mention these three in particular because I expect our code has many of these flaws.

For "input validation", it is feasible to write a static checker to look for an established pattern of input validation, or its absence.

For the two "leakage" issues, we should not expect a static checker soon enough. A membrane placed between ZCF and each contract can protect against these. It can prohibit property mutation across the boundary. Better, it can harden all objects on first contact with the membrane. It can also ensure that function proxies are not called with surprising `this` bindings.

**Comment by erights** [comment]
Attn @michaelfig @katelynsills

**Comment by katelynsills** [comment]
> ZCF must be fully defensive against all possible misbehavior by the contract.

Is this proven? Or do we have other options? It might be easier to assume that ZCF goes bad if the contract is malicious, and do some extra checks on the Zoe side. For instance, if we check offer safety on the Zoe side, we can be assured that no matter what happens to ZCF, offer safety holds. Payout liveness is already enforced on the Zoe side. What else needs to be enforced?

**Comment by erights** [comment]
Yes, it is possible to arrange to be less vulnerable to a corrupted zcf. Certainly offer safety could be enforced by redundant checks by zoe, which may even be a good idea anyway. On payout liveness, I actually don't remember how the responsibility split between zcf and the rest of zoe. But it does seem plausible that payout liveness could also be less vulnerable to a corrupted zcf.

Even with those possibilities, I still think zcf should be defensive against the contract, and a defensive membrane may be the lowest cost way to do this with significant confidence. Both strategies are worth exploring. Doing both may or may not give us greater confidence than just doing one or the other --- depending on the complexity cost.
